### PR TITLE
Added logic to parse Docker Registry URLs.

### PR DESCRIPTION
**Pull request by jpoler**
Ready for merge.

Signed-off-by: Jon Poler <jonathan.poler@apcera.com>

### CODE REVIEW

**Comment by jpoler** [comment]
@alextoombs Will you go over this briefly to discern if there are any glaring problems here?

File: docker/registry_url.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+package docker
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+// Mandatory fields for a valid RegistryURL:
```
Comment (alextoombs) [inline docker/registry_url.go]: Seems like this should be included with a function, or as package-level documentation; not quite where it is.

File: docker/registry_url.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+// Scheme, Host
+//
+// Possible formats:
```
Comment (alextoombs) [inline docker/registry_url.go]: Maybe link to the schema on Dockerhub, and be explicit that this is a v1 registry-conforming utility?

File: docker/registry_url.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+// <scheme>://<host>/<namespace>/<repo>
+// <scheme>://<host>/<namespace>/<repo>:<tag>
+// <scheme>://<host>:<port>
+// <scheme>://<host>:<port>/<repo>
+// <scheme>://<host>:<port>/<repo>:<tag>
+// <scheme>://<host>:<port>/<namespace>
+// <scheme>://<host>:<port>/<namespace>/<repo>
+// <scheme>://<host>:<port>/<namespace>/<repo>:<tag>
+
+type DockerRegistryURL struct {
```
Comment (alextoombs) [inline docker/registry_url.go]: Needs a comment.

File: docker/registry_url.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	Scheme    string
```
Comment (alextoombs) [inline docker/registry_url.go]: I'd recommend embedded a `net.Url` to cover e.g. `Scheme` and `Host` at least. But I just like embedding.

File: docker/registry_url.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	Host      string
+	Port      string
+	Namespace string
```
Comment (alextoombs) [inline docker/registry_url.go]: These docker-specific fields may need documentation.

File: docker/registry_url.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+	if u.Scheme == "" {
+		return registryURL, errors.New("Registry URL must provide a scheme")
+	} else if u.Host == "" {
+		return registryURL, errors.New("Registry URL must provide a host")
+	}
+
+	registryURL.Scheme = u.Scheme
+
+	splitHost := strings.Split(u.Host, ":")
```
Comment (alextoombs) [inline docker/registry_url.go]: Use `net.SplitHostPort`, please.

File: docker/registry_url.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		registryURL.Host = splitHost[0]
+	} else {
+		registryURL.Host, registryURL.Port = splitHost[0], splitHost[1]
+	}
+
+	if u.Path != "" {
+		// Remove leading and trailing slashes to avoid edge cases.
+		path := strings.Trim(u.Path, "/")
+		if strings.HasPrefix(path, ":") {
+			return registryURL, fmt.Errorf("Path %s cannot be made up of just a tag", path)
```
Comment (alextoombs) [inline docker/registry_url.go]: nit: `%q` instead of `%s`

File: docker/registry_url.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+		}
+		var rest string
```
Comment (alextoombs) [inline docker/registry_url.go]: You can always do `var rest, tag string`

File: docker/registry_url_test.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
+// Copyright 2014 Apcera Inc. All rights reserved.
+
+package docker
+
+import "testing"
+
+func TestParseDockerRegistryURL(t *testing.T) {
```
Comment (alextoombs) [inline docker/registry_url_test.go]: This style of test is a great opportunity to do a table-driven test. [Here's a good article](http://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go) about the pattern.

File: docker/registry_url.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+type DockerRegistryURL struct {
+	Scheme    string
+	Host      string
+	Port      string
+	Namespace string
+	Repo      string
+	Tag       string
+}
+
+func ParseDockerRegistryURL(s string) (*DockerRegistryURL, error) {
```
Comment (alextoombs) [inline docker/registry_url.go]: Needs a comment describing intent.

File: docker/registry_url.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	splitHost := strings.Split(u.Host, ":")
+	if len(splitHost) == 1 {
+		registryURL.Host = splitHost[0]
+	} else {
+		registryURL.Host, registryURL.Port = splitHost[0], splitHost[1]
+	}
+
+	if u.Path != "" {
+		// Remove leading and trailing slashes to avoid edge cases.
+		path := strings.Trim(u.Path, "/")
```
Comment (alextoombs) [inline docker/registry_url.go]: I'd prefer a different variable name so it's not confused with the `path` package.

File: docker/registry_url.go (hunk 10 above)
Comment (alextoombs) [inline docker/registry_url.go]: Also, this function is pretty confusing. Could we break it up into a few internal chunks, or have comments describing each function?

**Comment by jpoler** [comment]
@alextoombs Hopefully that is a little bit more readable. Let me know what you think. I will be changing the tests to table-driven tests as soon as I get a chance.

File: docker/registry_url.go [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+// <scheme>://<host>:<port>/<namespace>/<repo>
+// <scheme>://<host>:<port>/<namespace>/<repo>:<tag>
+type DockerRegistryURL struct {
+	// Scheme can be http or https.
+	Scheme string
+
+	// Host is a Fully Qualified Domain Name.
+	Host string
+
+	// Port is optinal, and may not be present.
```
Comment (alextoombs) [inline docker/registry_url.go]: s/`optinal`/`optional`

File: docker/registry_url.go [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+	Port string
+
+	// The repo name can be namespaced, but in V1 Docker
```
Comment (alextoombs) [inline docker/registry_url.go]: Start comment with field name, please.

Also, FWIW, no need to be so crazy with the linebreaks, but that's up to you. Nobody enforces 80 chars.

File: docker/registry_url.go [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+	// This field is optional when specifying Docker registry source
+	// whitelists.
+	Repo string
+
+	// Tag specifies a desired version of the docker image.
+	// for instance, to specify ubuntu 14.04, the tag is 14.04.
+	Tag string
+}
+
+// TODO(Jon P) Move this logic onto the DockerOrigin struct.
```
Comment (alextoombs) [inline docker/registry_url.go]: nit: comment is specific to Continuum while this is a public utility repo.

File: docker/registry_url.go [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+// that it conforms to the Docker V1 registry URL schema.
+func ParseDockerRegistryURL(s string) (*DockerRegistryURL, error) {
+	registryURL := &DockerRegistryURL{}
+	u, err := url.Parse(s)
+	if err != nil {
+		return registryURL, err
+	}
+
+	if u.Scheme == "" || u.Host == "" {
+		return registryURL, fmt.Errorf("Registry URL must provide a scheme and host: %s", s)
```
Comment (alextoombs) [inline docker/registry_url.go]: `%q`, not `%s`. User input, most likely.

File: docker/registry_url.go [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+// parsePath parses the Registry URL path (after <scheme>:://<host>:<port>)
+// into namespace, repo, and tag. All of these are optional parts of the
+// path.
+func (url *DockerRegistryURL) parsePath(path string) error {
+	if path == "" {
+		return nil
+	}
+	path, err := cleanPath(path)
```
Comment (alextoombs) [inline docker/registry_url.go]: I'd check out [path.Clean](https://golang.org/pkg/path/#Clean) from the standard library.

Also, if you end up using that package, I'd prefer if you made your variable names not overlap with the package name.

File: docker/registry_url.go [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+		return err
+	}
+
+	prefix, tag, err := parseTag(path)
+	if err != nil {
+		return err
+	}
+	url.Tag = tag
+
+	namespace, repo, err := validateNamespaceAndRepo(prefix)
```
Comment (alextoombs) [inline docker/registry_url.go]: nit: `parseNamespaceAndRepo`, not `validate*`.

File: docker/registry_url_test.go [hunk 18]
```diff
@@ -0,0 +1,192 @@
+// Copyright 2014 Apcera Inc. All rights reserved.
```
Comment (alextoombs) [inline docker/registry_url_test.go]: 2015

File: docker/registry_url.go [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+
+	return nil
+}
+
+// cleanPath removes leading and trailing forward slashes
+// and makes sure that the path does not only contain a tag.
+func cleanPath(path string) (string, error) {
+	path = strings.Trim(path, "/")
+	if strings.HasPrefix(path, ":") {
+		return "", fmt.Errorf("Path cannot be made up of just a tag: %s", path)
```
Comment (alextoombs) [inline docker/registry_url.go]: `%q`

File: docker/registry_url.go [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+// parseTag splits the Repository tag from the prefix.
+func parseTag(path string) (prefix, tag string, err error) {
+	splitPath := strings.Split(path, ":")
+	if len(splitPath) == 1 {
+		prefix = splitPath[0]
+	} else if len(splitPath) == 2 {
+		prefix, tag = splitPath[0], splitPath[1]
+	} else {
+		// Unlikely edge case but it doesn't hurt to test for it.
+		return "", "", fmt.Errorf("Path should not contain more than one colon: %s", path)
```
Comment (alextoombs) [inline docker/registry_url.go]: `%q`

File: docker/registry_url.go [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+// validateNamespaceAndRepo splits and validates a given Docker repo name.
+func validateNamespaceAndRepo(path string) (namespace, repo string, err error) {
+	splitPath := strings.Split(path, "/")
+	if len(splitPath) == 1 {
+		repo = splitPath[0]
+	} else if len(splitPath) == 2 {
+		namespace, repo = splitPath[0], splitPath[1]
+	}
+	if len(splitPath) > 2 {
+		return "", "", fmt.Errorf("url Path after <scheme>://<host> must only contain a <namespace>/<repo> before :<port>, got: %s", path)
```
Comment (alextoombs) [inline docker/registry_url.go]: I'd trim the verbosity of this error message some. Think about a user getting this at the command line.

And, `%q`

File: docker/registry_url.go [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+		prefix, tag = splitPath[0], splitPath[1]
+	} else {
+		// Unlikely edge case but it doesn't hurt to test for it.
+		return "", "", fmt.Errorf("Path should not contain more than one colon: %s", path)
+	}
+	return prefix, tag, nil
+}
+
+// validateNamespaceAndRepo splits and validates a given Docker repo name.
+func validateNamespaceAndRepo(path string) (namespace, repo string, err error) {
```
Comment (alextoombs) [inline docker/registry_url.go]: See above comment w.r.t. function name, please.

File: docker/registry_url.go [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+		namespace, repo = splitPath[0], splitPath[1]
+	}
+	if len(splitPath) > 2 {
+		return "", "", fmt.Errorf("url Path after <scheme>://<host> must only contain a <namespace>/<repo> before :<port>, got: %s", path)
+	}
+	return namespace, repo, nil
+}
+
+// OrderedPieces returns a slice of strings in the order of precedence
+// for enforcing policy on a Docker Registry. The URL is validated when it is
```
Comment (alextoombs) [inline docker/registry_url.go]: I'd omit comments about "enforcing policy"; just gives you a general-to-specific slice of pieces.

File: docker/registry_url_test.go [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+			&DockerRegistryURL{},
+		},
+	}
+
+	for i, val := range testValues {
+		result, err := ParseDockerRegistryURL(val.input)
+		if err != nil {
+			if val.expectedError == nil {
+				t.Fatalf("Case %d: Unexpected error while parsing struct", i)
+			}
```
Comment (alextoombs) [inline docker/registry_url_test.go]: nit: FWIW, I like doing `t.Errorf` in these tests because it lets you see ALL of the failures.

File: docker/registry_url.go (hunk 16 above)
Comment (jpoler) [inline docker/registry_url.go]: `path.Clean` wouldn't remove leading slashes, unfortunately. I'd also rather not allow use of `.` and `..` in URLs, because I don't know if that behavior would be at all meaningful in the context of Docker URLs. Opinions welcome here.

**Comment by jpoler** [comment]
@alextoombs I'd like to merge this if you find that agreeable.

File: docker/registry_url.go [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+// be passed to ParseFullDockerRegistryURL. If not, it will try to parse the
+// URL as if it is a (namespace/)*repo(:tag)?
+func ParseDockerRegistryURL(s string) (*DockerRegistryURL, error) {
+	registryURL, err := ParseFullDockerRegistryURL(s)
+	if err == nil {
+		return registryURL, nil
+	}
+	// String didn't parse but was supposed to be a full registry URL.
+	if strings.HasPrefix(s, "http") || strings.HasPrefix(s, "https") {
+		return &DockerRegistryURL{}, fmt.Errorf("Invalid full Docker registry URL: %s", err)
```
Comment (alextoombs) [inline docker/registry_url.go]: I'd much rather you return a `nil` here with the error, rather than a pointer to an initialized struct.

File: docker/registry_url.go [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+	}
+
+	registryURL = &DockerRegistryURL{}
+	err = registryURL.parsePath(s)
```
Comment (alextoombs) [inline docker/registry_url.go]: Again, down here, I'd much rather be explicit:

```
if err := registryURL.parsePath(s); err != nil {
   return nil, err
}
return registryURL, nil
```

File: docker/registry_url.go [hunk 27]
```diff
@@ -0,0 +1,267 @@
...
+	return registryURL, err
+}
+
+// ParseFullDockerRegistryURL validates an input string URL to make sure
+// that it conforms to the Docker V1 registry URL schema.
+func ParseFullDockerRegistryURL(s string) (*DockerRegistryURL, error) {
+	registryURL := &DockerRegistryURL{}
+	u, err := url.Parse(s)
+	if err != nil {
+		return registryURL, err
```
Comment (alextoombs) [inline docker/registry_url.go]: Same thing down here; explicit nil returns in `err != nil` cases, please.

File: docker/registry_url.go (hunk 23 above)
Comment (alextoombs) [inline docker/registry_url.go]: Same comment.

File: docker/registry_url.go (hunk 26 above)
Comment (jpoler) [inline docker/registry_url.go]: parsePath mutates the struct, so it needs to be fresh.
```
registryURL = &DockerRegistryURL{}
if err = registryURL.parsePath(s); err != nil {
	return nil, err
}
return registryURL, nil
```

Is that acceptable?

File: docker/registry.go [hunk 28]
```diff
@@ -15,7 +15,10 @@ import (
 
 var (
 	// DockerHubRegistryURL points to the official Docker registry.
-	DockerHubRegistryURL = "https://index.docker.io"
+	DockerHubRegistryURL    = "https://index.docker.io"
+	DockerHubRegistryScheme = "https"
+	DockerHubRegistryHost   = "index.docker.io"
```
Comment (alextoombs) [inline docker/registry.go]: Why do we have all of these others `vars`? Why are they exported?

File: docker/registry_url.go [hunk 29]
```diff
@@ -0,0 +1,268 @@
...
+//
+// Possible formats:
+//
+// <scheme>://[user:password@]<host>[:<port>][/<namespace>/<repo>[:<tag>]]
+type DockerRegistryURL struct {
+	// Scheme can be http or https.
+	Scheme string `json:",omitempty"`
+
+	// Userinfo holds basic auth credentials.
+	// NOTE: these MUST be removed from the DockerOrigin struct before being sent
```
Comment (alextoombs) [inline docker/registry_url.go]: This is specific to Continuum, and should be removed.

Comment (olegshaldybin) [inline docker/registry_url.go]: This comment is specific to how Continuum handles things, might not really belong in util package.

File: docker/registry_url.go [hunk 30]
```diff
@@ -0,0 +1,268 @@
...
+		return "", "", fmt.Errorf("URL Path must only contain a <namespace>/<repo>, got: %q", s)
+	}
+	return namespace, repo, nil
+}
+
+// OrderedPieces returns a slice of strings in the order of precedence in a URL
+// Scheme. The URL is validated when it is parsed so the only logic needed here
+// is to make sure that the order is some subset of the following sequence:
+// scheme -> host -> port -> namespace -> repo -> tag.
+func (u *DockerRegistryURL) OrderedPieces() []string {
```
Comment (olegshaldybin) [inline docker/registry_url.go]: What's the use case for OrderedPieces?

Comment (jpoler) [inline docker/registry_url.go]: https://github.com/apcera/continuum/blob/docker-source-policy-take2/common/pengine/ask/docker.go#L147-L172

It basically allows for iteration of the pieces from left to right for the length of each whitelist URL. If we are checking a "source" to see if it is allowed, I'm getting the ordered pieces for the "source" and the "rule". If any source piece does not match the rule piece at that index, the rule does not allow the source. The source must match the rule from left to right for the entire length of the rule for the source to be allowed.

**Comment by jpoler** [comment]
@olegshaldybin This needs to be merged before the continuum PR.

**Comment by olegshaldybin** [comment]
LGTM
